### PR TITLE
Update Validators

### PR DESCRIPTION
**Pull request by LeoKHoward**
# Motivation and Context
Confluence updated with more detailed validation/rules and took opportunity to fine tune current validation where possible

# What has changed
Postcode validator
Pipe character validator
CE E validator with two treatment code exceptions
Tests

# How to test?
Run make test
Validate a sample file

# Links
https://collaborate2.ons.gov.uk/confluence/pages/viewpage.action?spaceKey=CRM&title=Address+Frame+specification+2021
https://collaborate2.ons.gov.uk/confluence/pages/viewpage.action?spaceKey=CRM&title=ESTAB_TYPES+2021
https://trello.com/c/MmDZrIjk/1245-update-validation-on-sample-load-and-bulk-new-address-8

### CODE REVIEW

**Comment by LeoKHoward** [comment]
[Update validation on sample load and bulk new address (8)](https://trello.com/c/MmDZrIjk/1245-update-validation-on-sample-load-and-bulk-new-address-8)

File: validators.py [hunk 1]
```diff
@@ -82,10 +82,12 @@ def validate(value: Iterable, **_kwargs) -> None:
     return validate
 
 
-def no_padding_whitespace():
+def no_padding_whitespace_and_no_pipe_character():
```
Comment (AdamHawtin) [inline validators.py]: I think these should be separate validators, like this if a value had padding whitespace and a pipe, the pipe won't be spotted.

Also the check for the pipe will be relatively slow over lots of rows, we can be savvy with which columns we apply it too (e.g. column that uses the `in_set` or `numeric` validator doesn't need to also check for the pipe explicitly).

**Review by AdamHawtin — DISMISSED** [review]

**Review by NickGrantX — CHANGES_REQUESTED** [review]
Things I've found (repeated on both cards, because they have the code duplicated):
- Estab type doesn't need a max - already restricted to set
- Llanfairpwllgwyngyllgogerychwyrndrobwllllantysiliogogogoch is 58 characters, but town name is max 30
- Lat/long should be between -180.0 to +180.0 (in fact probably not more than 14 degrees West to be honest... etc see https://en.wikipedia.org/wiki/List_of_extreme_points_of_the_United_Kingdom#:~:text=Westernmost%20point%20%E2%80%93%20Rockall%20at%2057,W%20is%20the%20westernmost%20point.)
- HTC_WILLINGNESS, HTC_DIGITAL, TREATMENT_CODE - don't check for padding or whitespace, because already restricted to set

File: validate_sample.py [hunk 2]
```diff
@@ -18,41 +19,47 @@ class SampleValidator:
...
 
     def __init__(self):
         self.schema = {
-            'UPRN': [mandatory(), max_length(12), numeric(), no_padding_whitespace()],
-            'ESTAB_UPRN': [mandatory(), max_length(12), numeric(), mandatory(), no_padding_whitespace()],
+            'UPRN': [mandatory(), max_length(13), numeric(), no_padding_whitespace()],
+            'ESTAB_UPRN': [mandatory(), max_length(13), numeric(), no_padding_whitespace()],
             'ADDRESS_TYPE': [mandatory(), in_set({'HH', 'CE', 'SPG'}), no_padding_whitespace()],
-            'ESTAB_TYPE': [mandatory(), in_set(self.ESTAB_TYPES), max_length(255), no_padding_whitespace()],
+            'ESTAB_TYPE': [mandatory(), in_set(self.ESTAB_TYPES), no_padding_whitespace()],
```
Comment (NickGrantX) [inline validate_sample.py]: doesn't need to check for padding/whitespace because it already has an `in_set` check

File: validate_sample.py [hunk 3]
```diff
@@ -18,41 +19,47 @@ class SampleValidator:
...
             'ADDRESS_LEVEL': [mandatory(), in_set({'E', 'U'}), no_padding_whitespace()],
```
Comment (NickGrantX) [inline validate_sample.py:36]: doesn't need to check for padding/whitespace because it already has an `in_set` check

**Review by NickGrantX — CHANGES_REQUESTED** [review]
`latitude_longitude` still doesn't check to see if the location is on the planet, at least, which is done by simply checking value is within range -180 to 180

**Review by NickGrantX — APPROVED** [review]
I would have done lat-long validation in a single validator function, but no reason to block the PR

File: validators.py [hunk 4]
```diff
@@ -107,3 +115,32 @@ def validate(expected_capacity, **kwargs):
...
+        stripped_postcode = postcode.replace(" ", "")
+        if not stripped_postcode.isalnum():
+            raise Invalid(f'Postcode "{postcode}" is non alphanumeric')
+
+    return validate
+
+
+def latitude_longitude_range():
+    def validate(value, **_kwargs):
+        if not -180 <= float(value) <= 180:
```
Comment (AdamHawtin) [inline validators.py]: This needs to catch a `ValueError` and raise it as an `Invalid`, as is if the value can't be converted to float it kills the script.

**Review by ryangrundy7 — APPROVED** [review]

**Pull request merged**